File: static/js/visualizers/circle_spectrum.js
// =============================================================================
// visualizers/circle_spectrum.js
//
// サークルスペクトログラム GL プラグイン。Python 版
// plugins/visualizers/circle_spectrum.py の出力を canvas2d + CanvasTexture で
// 再現する。
//
// 戦略:
//   - サーバ側 gl_data_streams が per-frame の生 dB スペクトル (Float32) を
//     `spectrum` ストリームとして書き出す。shape = [frameCount, bands]。
//   - 各フレームで該当 bands 行を切り出し、(dbFloor..dbCeil) で [0..1] へ正規化、
//     中心 (cx, cy) から放射状にバーを引く。
//   - dbFloor / dbCeil / rotateDeg / mirror 等の見た目パラメータはブラウザ側で
//     扱うため、UI で値が変わってトークン再ベイクが走った直後も「ベイクされた
//     spectrum 行 × ブラウザ側の最新閾値」で表示が一致する。
// =============================================================================

const TAU = Math.PI * 2;

function numParam(value, fallback) {
  const raw = value ?? fallback;
  const n = Number(raw);
  return Number.isFinite(n) ? n : fallback;
}

function normalizeParams(rawParams, width, height) {
  const p = rawParams || {};
  let dbFloor = numParam(p.dbFloor, -75);
  let dbCeil = numParam(p.dbCeil, -20);
  if (dbCeil <= dbFloor) dbCeil = dbFloor + 1;
  return {
    color: String(p.color || "#ffffff"),
    cx: numParam(p.centerX, width / 2),
    cy: numParam(p.centerY, height / 2),
    inner: Math.max(0, numParam(p.innerRadius, 220)),
    barMax: Math.max(1, numParam(p.barLength, 180)),
    lineWidth: Math.max(1, numParam(p.lineWidth, 4)),
    bands: Math.max(8, numParam(p.bands, 96)),
    dbFloor,
    dbCeil,
    rotateDeg: numParam(p.rotateDeg, -90),
    mirror: String(p.mirror || "outer"),
    opacity: Math.max(0, Math.min(1, numParam(p.opacity, 0.95))),
  };
}

function paramsCacheKey(p) {
  return [
    p.color, p.cx, p.cy, p.inner, p.barMax, p.lineWidth, p.bands,
    p.dbFloor, p.dbCeil, p.rotateDeg, p.mirror, p.opacity,
  ].join("|");
}

export async function createVisualizerLayer(ctx) {
  const { THREE, width, height, params, audioData, streamShapes, frameCount } = ctx;

  const canvas = document.createElement("canvas");
  canvas.width = Math.max(1, Math.round(width));
  canvas.height = Math.max(1, Math.round(height));
  const c2d = canvas.getContext("2d");

  const texture = new THREE.CanvasTexture(canvas);
  texture.colorSpace = THREE.SRGBColorSpace;
  texture.flipY = false;
  texture.minFilter = THREE.LinearFilter;
  texture.magFilter = THREE.LinearFilter;
  texture.generateMipmaps = false;

  const geom = new THREE.PlaneGeometry(canvas.width, canvas.height);
  const mat = new THREE.MeshBasicMaterial({
    map: texture,
    transparent: true,
    depthTest: false,
    depthWrite: false,
    toneMapped: false,
    side: THREE.DoubleSide,
    forceSinglePass: true,
  });
  const mesh = new THREE.Mesh(geom, mat);
  mesh.position.set(canvas.width / 2, canvas.height / 2, 0);
  mesh.frustumCulled = false;

  const specFlat = audioData?.spectrum || null;
  const specShape = streamShapes?.spectrum || null;
  // shape = [frameCount, bands]。バンド数はサーバが gl_data_streams で焼いた値が
  // 正で、UI の bands パラメータと一致する。フォールバックは flat 全長 / frameCount。
  const bandsBaked = (() => {
    if (Array.isArray(specShape) && specShape.length >= 2) return Math.max(1, Number(specShape[1]));
    if (specFlat && frameCount > 0) return Math.max(1, Math.floor(specFlat.length / frameCount));
    return 0;
  })();

  let lastFrameIdx = -1;
  let lastParamsKey = "";

  function drawFallbackCircle(p) {
    // 音源未指定時の表示。Python 版 (ctx.audio is None) と揃えて細い円だけ描く。
    c2d.save();
    c2d.globalAlpha = Math.max(0, Math.min(1, p.opacity * (60 / 255)));
    c2d.strokeStyle = p.color;
    c2d.lineWidth = Math.max(1, Math.floor(p.lineWidth / 2));
    c2d.beginPath();
    c2d.arc(p.cx, p.cy, p.inner, 0, TAU);
    c2d.stroke();
    c2d.restore();
  }

  function drawSpectrum(p, frameIdx) {
    if (!specFlat || bandsBaked <= 0) return;
    const off = frameIdx * bandsBaked;
    if (off + bandsBaked > specFlat.length) return;
    // バンド数 (UI 値) がベイク値と異なる場合は線形リサンプル。通常は一致する
    // (token に bands が含まれるので変更時は再ベイク)。
    const bandsOut = Math.max(1, Math.round(p.bands));
    const baseAngle = (p.rotateDeg * Math.PI) / 180;
    const range = p.dbCeil - p.dbFloor;

    c2d.save();
    c2d.globalAlpha = p.opacity;
    c2d.strokeStyle = p.color;
    c2d.lineWidth = p.lineWidth;
    c2d.lineCap = "butt";

    for (let i = 0; i < bandsOut; i++) {
      let dbValue;
      if (bandsOut === bandsBaked) {
        dbValue = specFlat[off + i];
      } else {
        // 線形補間。UI スライダで bands を変える → 再ベイク待ちの瞬間の見た目が
        // 多少崩れても良い、という割り切り。
        const f = (i / Math.max(1, bandsOut - 1)) * (bandsBaked - 1);
        const lo = Math.floor(f);
        const hi = Math.min(bandsBaked - 1, lo + 1);
        const t = f - lo;
        dbValue = specFlat[off + lo] * (1 - t) + specFlat[off + hi] * t;
      }
      let v = (dbValue - p.dbFloor) / range;
      if (v <= 0.001) continue;
      if (v > 1) v = 1;
      const a = baseAngle + (i / bandsOut) * TAU;
      const cosA = Math.cos(a);
      const sinA = Math.sin(a);
      const barLen = v * p.barMax;
      const x1 = p.cx + cosA * p.inner;
      const y1 = p.cy + sinA * p.inner;
      const x2 = p.cx + cosA * (p.inner + barLen);
      const y2 = p.cy + sinA * (p.inner + barLen);
      c2d.beginPath();
      c2d.moveTo(x1, y1);
      c2d.lineTo(x2, y2);
      c2d.stroke();
      if (p.mirror === "both") {
        const innerBack = Math.max(0, p.inner - barLen);
        const xb = p.cx + cosA * innerBack;
        const yb = p.cy + sinA * innerBack;
        c2d.beginPath();
        c2d.moveTo(x1, y1);
        c2d.lineTo(xb, yb);
        c2d.stroke();
      }
    }
    c2d.restore();
  }

  function renderInto(p, frameIdx) {
    c2d.clearRect(0, 0, canvas.width, canvas.height);
    if (p.opacity <= 0) return;
    if (!specFlat || frameCount <= 0 || bandsBaked <= 0) {
      drawFallbackCircle(p);
      return;
    }
    drawSpectrum(p, frameIdx);
  }

  function updateImpl(frameIdx) {
    const p = normalizeParams(params, canvas.width, canvas.height);
    const pk = paramsCacheKey(p);
    if (frameIdx === lastFrameIdx && pk === lastParamsKey) return;
    lastFrameIdx = frameIdx;
    lastParamsKey = pk;
    renderInto(p, frameIdx);
    texture.needsUpdate = true;
  }

  updateImpl(0);

  function update(frameState) {
    const idx = Math.max(0, Math.floor(Number(frameState?.frameIdx) || 0));
    updateImpl(idx);
  }
  function dispose() {
    geom.dispose();
    mat.dispose();
    texture.dispose();
  }

  return { object3D: mesh, update, dispose };
}

File: plugins/visualizers/circle_spectrum.py
"""サークルスペクトログラム。

中心 (centerX, centerY) を基準に、円周上に等間隔でスペクトルバンドのバーを
外向き (または両側) に伸ばす。dB → バー長は browser 側 (``dbFloor / dbCeil``) で正規化。

GL plugin: サーバ側で per-frame の生 dB スペクトル (FFT 後 log ビン化したもの) を
Float32 binary stream として書き出し、ブラウザの canvas2d で円周バーを描画する。
"""
from __future__ import annotations

import numpy as np

KEY = "circle_spectrum"
NAME = "サークルスペクトログラム"

GL_MODULE = "/static/js/visualizers/circle_spectrum.js"
GL_VERSION = 1

# 解析に影響するのは FFT 系のみ。dbFloor/dbCeil はブラウザ側正規化なので含めない。
ANALYSIS_KEYS = ["bands", "fmin", "fmax", "windowMs"]

PARAMS = [
    {"key": "color", "type": "color", "default": "#ffffff", "label": "色"},
    {"key": "centerX", "type": "number", "min": 0, "max": 1920, "step": 10, "default": 960, "label": "中心 X"},
    {"key": "centerY", "type": "number", "min": 0, "max": 1080, "step": 10, "default": 540, "label": "中心 Y"},
    {"key": "innerRadius", "type": "number", "min": 0, "max": 800, "step": 10, "default": 220, "label": "内側半径 (px)"},
    {"key": "barLength", "type": "number", "min": 10, "max": 800, "step": 10, "default": 180, "label": "最大バー長 (px)"},
    {"key": "lineWidth", "type": "number", "min": 1, "max": 32, "step": 1, "default": 4, "label": "線の太さ (px)"},
    {"key": "bands", "type": "number", "min": 16, "max": 256, "step": 1, "default": 96, "label": "バンド数"},
    {"key": "fmin", "type": "number", "min": 20, "max": 2000, "step": 10, "default": 60, "label": "下限 Hz"},
    {"key": "fmax", "type": "number", "min": 1000, "max": 22000, "step": 100, "default": 12000, "label": "上限 Hz"},
    {"key": "windowMs", "type": "number", "min": 10, "max": 200, "step": 5, "default": 50, "label": "FFT 窓 (ms)"},
    {"key": "dbFloor", "type": "number", "min": -120, "max": -10, "step": 1, "default": -75, "label": "下限 dB"},
    {"key": "dbCeil", "type": "number", "min": -60, "max": 0, "step": 1, "default": -20, "label": "上限 dB"},
    {"key": "rotateDeg", "type": "number", "min": -360, "max": 360, "step": 1, "default": -90, "label": "回転 (°)"},
    {"key": "mirror", "type": "select", "options": [
        {"value": "outer", "label": "外側のみ"},
        {"value": "both", "label": "内外両側"},
    ], "default": "outer", "label": "向き"},
    {"key": "opacity", "type": "number", "min": 0, "max": 1, "step": 0.05, "default": 0.95, "label": "不透明度"},
]


def gl_data_streams(params, audio, time_grid_sec, fps):
    """GL 経路向け per-frame スペクトル (dB) を返す。

    各フレーム (time_grid_sec[i]) を右端とする windowMs 長の窓に Hann + rfft を
    かけ、log ビン化した dB スペクトル (`AudioContext.spectrum_db` と同じ式)
    を ``spectrum`` ストリームとして書き出す。dbFloor / dbCeil の正規化は
    ブラウザ側で実施 (UI で値を動かしたときも再ベイクされて反映される)。

    返却 shape:
      ``spectrum`` -> ``(N_frames, bands)``  float32 dBFS (おおむね -120..0)
    """
    grid = np.asarray(time_grid_sec, dtype=np.float64)
    n_frames = int(grid.size)
    try:
        bands = max(8, int(params.get("bands", 96)))
    except (TypeError, ValueError):
        bands = 96
    if audio is None or n_frames <= 0:
        return {"spectrum": np.full((max(0, n_frames), bands), -120.0, dtype=np.float32)}

    try:
        window_sec = max(0.010, float(params.get("windowMs") or 50) / 1000.0)
    except (TypeError, ValueError):
        window_sec = 0.050
    try:
        fmin = max(1.0, float(params.get("fmin") or 60))
        fmax = max(fmin + 1.0, float(params.get("fmax") or 12000))
    except (TypeError, ValueError):
        fmin, fmax = 60.0, 12000.0

    out = audio.batch_spectrum_db(
        grid,
        n_bands=bands,
        window_sec=window_sec,
        fmin=fmin,
        fmax=fmax,
    )
    return {"spectrum": out}
